Assets/Scripts/Algorithms/Patrolling/Components/VirtualStigmergyComponent.cs

```diff
@@ -0,0 +1,353 @@
+
+            return null;
+
+        }
+
+        /// <summary>
+        /// Gets a value in the stigmergy.
+        /// </summary>
+        /// <param name="key">The key to get the value for.</param>
+        /// <returns>The value or null if key is not present.</returns>
+        /// <remarks>This only returns what is in the local knowledge, but it sends a get message to receive the newest information from neighbors. This information is first available next tick, however.</remarks>
+        public string? Get(string key)
+        {
+            if (!_localKnowledge.TryGetValue(key, out var valueInfo))
+            {
+                // Ask neighbors for this information.
+                // It won't be immediately available, so we won't be able to return it here.
+                // NOTE: It is unclear what the timestamp should be if we have nothing in our local knowledge.
+                // So i will give it 0, it should always be smaller than any timestamp of any existing entries.
+                BroadcastMessage(VirtualStigmergyMessage.CreateGetMessage(key, null, 0, _monaRobot.id));
+
+                return null;
+            }
+
+            // Ask neighbors if our information is up to date.
+            // It won't be immediately available, so we won't be able to return it here.
+            BroadcastMessage(VirtualStigmergyMessage.CreateGetMessage(key, valueInfo.Value, valueInfo.Timestamp, valueInfo.RobotId));
+
+            return valueInfo.Value;
+        }
+
+        /// <summary>
+        /// Checks if a key exists in the local knowledge
+        /// </summary>
+        /// <param name="key">The key to check existance for.</param>
+        /// <returns><see langword="true"/> if the key exists otherwise <see langword="false"/></returns>
+        /// <remarks>This does not communicate with neighbors.</remarks>
+        public bool Has(string key)
+        {
+            // The paper is a little vague what Has does.
+            // Should it ask the swarm like Get? or should it just check the local knowledge?
+            // I have implemented it using the local knowledge just as Size.
+            return _localKnowledge.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// Gets how many key-value pairs are in the local knowledge.
+        /// </summary>
+        /// <returns></returns>
+        /// <remarks>This does not communicate with neighbors.</remarks>
+        public int Size()
+        {
+            return _localKnowledge.Count;
+        }
+
+        private void BroadcastMessage(VirtualStigmergyMessage message)
+        {
+#if VIRTUAL_STIGMERGY_TRACING
+            Debug.LogFormat("STIGMERGY Robot {0} sending {1} message: key: {2}, value: {3}, timestamp: {4}, robotId: {5}", _monaRobot.id, message.Type, message.Key, message.Value, message.Timestamp, message.RobotId);
+#endif
+            _communicationManager.BroadcastMessage(_monaRobot, message);
+        }
+
+        public readonly struct ValueInfo
```

Comment:
Should we make the ValueInfo generic, such that the Value can be anything than string?

Reply:
done